src/Streamly/Internal/Data/Producer.hs

```diff
@@ -0,0 +1,235 @@
+-- |
+-- Module      : Streamly.Internal.Data.Producer
+-- Copyright   : (c) 2021 Composewell Technologies
+-- License     : BSD-3-Clause
+-- Maintainer  : streamly@composewell.com
+-- Stability   : experimental
+-- Portability : GHC
+--
+-- A 'Producer' is an 'Unfold' with an 'extract' function added to to extract
+-- the state. It is more powerful but less general than an Unfold.
+--
+-- A 'Producer' represents steps of a loop generating a sequence of elements.
+-- While unfolds are closed representation of imperative loops with some opaque
+-- internal state, producers are open loops with the state being accesible to
+-- the user.
+--
+-- Unlike an unfold, which runs a loop till completion, a producer can be
+-- stopped in the middle, its state can be extracted, examined, changed, and
+-- then it can be resumed later from the stopped state.
+--
+-- A producer can be used in places where a CPS stream would otherwise be
+-- needed, because the state of the loop can be passed around. However, it can
+-- be much more efficient than CPS because it allows stream fusion and
+-- unecessary function calls can be avoided.
+
+module Streamly.Internal.Data.Producer
+    ( Unfold (..)
+
+    -- * Converting
+    , simplify
+
+    -- * Unfolds
+    , nil
+    , nilM
+    , unfoldrM
+    , fromStreamD
+    , fromList
+
+    -- * Combinators
+    , NestedLoop (..)
+    , concat
+    , parse
+    , parseMany
+    , parseManyD
+    )
+where
+
+#include "inline.hs"
+
+import Control.Exception (assert)
+import Control.Monad.Catch (MonadThrow, throwM)
+import GHC.Exts (SpecConstrAnnotation(..))
+import GHC.Types (SPEC(..))
+import Streamly.Internal.Data.Parser.ParserD (ParseError(..), Step(..))
+import Streamly.Internal.Data.Producer.Source (Source)
+import Streamly.Internal.Data.Stream.StreamD.Step (Step(..))
+import Streamly.Internal.Data.Stream.StreamD.Type (Stream(..))
+import Streamly.Internal.Data.SVar (defState)
+import Streamly.Internal.Data.Unfold.Types (Unfold(..))
+
+import qualified Streamly.Internal.Data.Parser.ParserD as ParserD
+import qualified Streamly.Internal.Data.Parser.ParserK.Types as ParserK
+import qualified Streamly.Internal.Data.Producer.Source as Source
+
+import Streamly.Internal.Data.Producer.Type
+import Prelude hiding (concat)
+
+-- XXX We should write unfolds as producers where possible and define
```

Comment:
I don't think we need to define unfolds using `simplify`. The users can do that themselves if ever need (They should ideally use producers where they can)